cordapp/src/main/kotlin/com/r3/corda/finance/obligation/client/flows/UpdatePaymentStatus.kt

```diff
@@ -0,0 +1,105 @@
+import net.corda.core.contracts.UniqueIdentifier
+import net.corda.core.flows.*
+import net.corda.core.transactions.SignedTransaction
+import net.corda.core.transactions.TransactionBuilder
+import net.corda.core.transactions.WireTransaction
+import net.corda.core.utilities.ProgressTracker
+
+object UpdatePaymentStatus {
+
+    @InitiatingFlow
+    @StartableByRPC
+    class Initiator(
+            private val paymentReference: PaymentReference,
+            private val status: PaymentStatus,
+            private val linearId: UniqueIdentifier
+    ) : FlowLogic<WireTransaction>() {
+
+        companion object {
+            object INITIALISING : ProgressTracker.Step("Performing initial steps.")
+            object BUILDING : ProgressTracker.Step("Building and verifying transaction.")
+            object SIGNING : ProgressTracker.Step("signing transaction.")
+
+            object COLLECTING : ProgressTracker.Step("Collecting counterparty signature.") {
+                override fun childProgressTracker() = CollectSignaturesFlow.tracker()
+            }
+
+            object FINALISING : ProgressTracker.Step("Finalising transaction.") {
+                override fun childProgressTracker() = FinalityFlow.tracker()
+            }
+
+            fun tracker() = ProgressTracker(INITIALISING, BUILDING, SIGNING, COLLECTING, FINALISING)
+        }
+
+        override val progressTracker: ProgressTracker = tracker()
+
+        @Suspendable
+        override fun call(): WireTransaction {
+            progressTracker.currentStep = INITIALISING
+            val obligationStateAndRef = getLinearStateById<Obligation<*>>(linearId, serviceHub)
+                    ?: throw IllegalArgumentException("LinearId not recognised.")
+            val obligationState = obligationStateAndRef.state.data
+
+            val os = obligationState.withWellKnownIdentities(resolver)
+            val obligee = os.obligee
+            check(ourIdentity == obligee) { "This flow can only be started by the obligee. " }
+            val obligor = serviceHub.identityService.requireWellKnownPartyFromAnonymous(os.obligor)
+            val payment = obligationState.payments.find { it.paymentReference == paymentReference }
+            check(payment != null) { "Could not find payment with reference '$paymentReference'" }
+
+            progressTracker.currentStep = BUILDING
+            payment!!.status = status
+            val notary = serviceHub.networkMapCache.notaryIdentities.firstOrNull()
+                    ?: throw FlowException("No available notary.")
+            val utx = TransactionBuilder(notary = notary).apply {
+                addInputState(obligationStateAndRef)
+                val signers = obligationState.participants.map { it.owningKey }
+                addCommand(ObligationCommands.UpdatePayment(payment.paymentReference), signers)
+                addOutputState(obligationState, ObligationContract.CONTRACT_REF)
```

Comment:
is the status being updated here? Looks like you are just adding back the input obligation.

Reply:
The status has been changed on line 61 because Payment is mutable

Reply:
Ah yes, thanks!